File: src/patterns/creational/abstractfactory/abstractfactory.nim
# internal
import
    patterns/creational/abstractfactory/animal


type
    AbstactFactory*[TAnimal: Animal] = ref object
        animal: TAnimal


proc newAbstactFactory*[T: Animal](animal: T): AbstactFactory[T] =
    return AbstactFactory[T](animal: animal)

proc getAnimalInfo*(self: AbstactFactory): string =
    return "Name: " & self.animal.getName() & "\nSpecie Name: " & self.animal.getSpecieName()

File: src/patterns/creational/abstractfactory/abstractfactory.go
package abstractfactory


type AbstractFactory struct {
	animal Animal
}


func NewAbstactFactory(animal Animal) *AbstractFactory {
	return &AbstractFactory{
		animal: animal,
	}
}

func (self *AbstractFactory) GetAnimalInfo() string {
	return "Name: " + self.animal.GetName() + "\nSpecie Name: " + self.animal.GetSpecieName()
}